crates/amaru/src/stages/consensus/forward_chain/tcp_forward_chain_server.rs

```diff
@@ -0,0 +1,211 @@
+            })
+            .me;
+
+        let our_tip = Arc::new(Mutex::new(our_tip.clone()));
+        runtime.with_rt(|rt| {
+            let our_tip_clone = our_tip.clone();
+            let clients_clone = clients.clone();
+            rt.spawn(async move {
+                loop {
+                    // due to the signature of TcpListener::accept, this is the only way to use this API
+                    // in particular, it isn’t possible to poll for new peers within the `schedule` method
+                    match PeerServer::accept(&tcp_listener, network_magic).await {
+                        Ok(peer) => {
+                            let our_tip = our_tip_clone.lock().unwrap().clone();
+                            clients_clone.send(ClientMsg::Peer(
+                                peer,
+                                Tip(our_tip.point().pallas_point(), our_tip.block_height()),
+                            ));
+                        }
+                        Err(e) => {
+                            tracing::warn!(
+                                target: EVENT_TARGET,
+                                "error accepting peer: {e}",
+                            );
+                            continue;
+                        }
+                    };
+                }
+            })
+        });
+
+        Ok(Self {
+            _runtime: runtime,
+            clients: clients.clone(),
+            our_tip: our_tip.clone(),
+        })
+    }
+}
+
+/// This implementation of ForwardEventListener sends the received events to all connected clients.
+#[async_trait]
+impl ForwardEventListener for TcpForwardChainServer {
+    async fn send(&self, event: ForwardEvent) -> anyhow::Result<()> {
+        match event {
+            ForwardEvent::Forward(header) => {
+                let mut our_tip = self
+                    .our_tip
+                    .lock()
+                    .map_err(|e| anyhow::anyhow!("Mutex poisoned: {}", e))?;
+                *our_tip = header.as_header_tip();
+                self.clients.send(ClientMsg::Op(ClientOp::Forward(header)));
+                Ok(())
+            }
+            ForwardEvent::Backward(tip) => {
+                let mut our_tip = self
+                    .our_tip
+                    .lock()
+                    .map_err(|e| anyhow::anyhow!("Mutex poisoned: {}", e))?;
+                *our_tip = tip.clone();
+                self.clients.send(ClientMsg::Op(ClientOp::Backward(Tip(
+                    tip.point().pallas_point(),
+                    tip.block_height(),
+                ))));
+                Ok(())
+            }
+        }
+    }
+}
+
+async fn client_supervisor(
+    mut cell: ActoCell<ClientMsg, impl ActoRuntime, anyhow::Result<()>>,
+    store: Arc<dyn ChainStore<Header>>,
+    max_peers: usize,
+) {
+    let mut clients = BTreeMap::new();
+    while let Some(msg) = cell.recv().await.has_senders() {
```

Comment:
`has_senders` will always remain true because of the ActoRef held in the listener task. If we want to shut this down, we’ll need a dedicated ClientMsg for that purpose.